### PR TITLE
printchplbuilds: displays information about available runtime builds

### PR DESCRIPTION
**Pull request by jhh67**
`printchplbuilds` is a tool that displays information about the available
runtime builds. It is helpful when `chpl` complains that a runtime for
your current configuration is not available. It has options for
sorting the builds, changing how the builds are displayed, emitting
shell commands to match a particular build, and checking that a build
exists that matches the current configuration.

Signed-off-by: John H. Hartman <jhh67@users.noreply.github.com>

### CODE REVIEW

File: util/chplenv/printchplbuilds.py [hunk 1]
```diff
@@ -0,0 +1,449 @@
...
+            diffvars |= set(FindDiffVars(usedvars, current, builds[i]))
+        values = {}
+        for config in [current] + builds:
+            for var in usedvars:
+                if var in diffvars:
+                    value = config.get(var, 'NA')
+                    if var not in values.keys():
+                        values[var] = set([value])
+                    else:
+                        values[var].add(value)
```
Comment (aconsroe-hpe) [inline util/chplenv/printchplbuilds.py:371]: values can be a `defaultdict(set)` here and then do `values[var].add(value)` for a minor cleanup

File: util/chplenv/printchplbuilds.py [hunk 2]
```diff
@@ -0,0 +1,449 @@
...
+            for config in configs[done:done+columns]:
+                value = config.get(var, 'NA')
+                if var not in extras:
+                    tag = "" if current.get(var, 'NA') == value else "*"
+                    value += tag
+                if var == 'MTIME' and value != 'NA':
+                    value = datetime.datetime.fromtimestamp(value).strftime("%b %d %H:%M")
+                output += " %-20s" % value
+            print(output)
+        print
```
Comment (aconsroe-hpe) [inline util/chplenv/printchplbuilds.py:444]: leftover?

File: util/chplenv/printchplbuilds.py [hunk 3]
```diff
@@ -0,0 +1,449 @@
...
+STATE_NONE = 0
+STATE_TARGET_PLATFORM = 1
+STATE_TARGET_COMPILER = 2
+STATE_PREFIX = 3
+STATE_HWLOC = 4
+STATE_RE2 = 5
+STATE_LIBFABRIC = 6
+STATE_TARGET_ARCH = 7
+STATE_COMM_SUBSTRATE = 8
+STATE_GASNET_SEGMENT = 9
```
Comment (aconsroe-hpe) [inline util/chplenv/printchplbuilds.py:32]: could use an Enum here

File: util/chplenv/printchplbuilds.py [hunk 4]
```diff
@@ -0,0 +1,449 @@
...
+# that correspond to that path. 
+
+import os
+import utils
+import printchplenv
+import sys
+import stat
+from collections import OrderedDict
+import optparse
+from pprint import pprint
```
Comment (aconsroe-hpe) [inline util/chplenv/printchplbuilds.py:13]: unused?

File: util/chplenv/printchplbuilds.py [hunk 5]
```diff
@@ -0,0 +1,449 @@
...
+    global configs
+    global used
+    global usedvars
+    usage = "Usage %prog [options]"
+    description = "Displays the available Chapel runtime builds. Values that differ from" \
+    " the current configuration have a '*' suffix. '+' denotes a variable that is set but" \
+    " whose value doesn't matter, '-' denotes a variable that is not set. 'NA' denotes a" \
+    " variable that is not applicable to the build." 
+
+    parser = optparse.OptionParser(version = "1.0", usage=usage, description=description)
```
Comment (aconsroe-hpe) [inline util/chplenv/printchplbuilds.py:220]: optparse has been deprecated for a long time now, but I know we already use it in quite a few places. Maybe worth using `argparse` for less future work? Or just as easy to do a single sweep over the whole repo if/when that becomes an issue. Not sure

**Review by aconsroe-hpe — APPROVED** [review]
Looks good I think this will be useful to get in. I made some comments on a few places to look at.

I am a bit scared this is prone to breakage when something in the build system changes. I wonder if printchplenv could provide something instead of having to parse the paths like you do. Maybe save a file like `thischplenv` in the build folder so you (and other tools) can just read that file to get all the information instead? Future work though I think

File: util/chplenv/printchplbuilds.py [hunk 6]
```diff
@@ -0,0 +1,443 @@
...
+                value = datetime.datetime.fromtimestamp(value).strftime("%b %d %H:%M")
+            output += " %-20s" % value
+        print(output)
+
+# For argument parsing
+@unique
+class Sort(Enum):
+    NEWEST =    1,
+    OLDEST =    2,
+    FEWEST =    3,
```
Comment (aconsroe-hpe) [inline util/chplenv/printchplbuilds.py:224]: I think the trailing comma here makes them tuples. Shouldn't actually cause an issue if you only ever compare by value but probably not intentional

**Comment by jhh67** [comment]
I had to abandon this PR because I had to delete the underlying branch. I had a couple of commits that were missing the DCA and while following git's instructions on how to fix the problem, during the rebase I ran into a patch that could not be applied. Rather than make the problem worse I created a new branch and applied the changes  manually which caused this PR to be closed. The PR that replaced this one is #18986.

**Pull request closed without merging**